### PR TITLE
sparse_flash_attention operator add to NPU kernel

### PR DESCRIPTION
**Pull request by AndyLi429**
add a sparse flash attention operators for sglang\npu.
## Unit Test
python tests/python/sgl_kernel_npu/test_sparse_flash_attention.py
test results:
Ran 2 tests in 0.970s

OK

### *eagle mode:*
python test_gms8k.py 
start
100/100 [03:36<00:00,  2.16s/it]
Accuracy: 0.980
Invalid: 0.000
Latency: 218.359 s
Output throughput: 40.681 token/s
metrics={'accuracy': 0.98, 'invalid': 0.0, 'latency': 218.35866242006887, 'output_throughput': 40.68077676218437}
metrics['accuracy']=0.98

### *Graph mode:*
python test_gms8k.py 
start
100/100 [02:17<00:00,  1.38s/it]
Accuracy: 0.980
Invalid: 0.000
Latency: 138.304 s
Output throughput: 62.175 token/s
metrics={'accuracy': 0.98, 'invalid': 0.0, 'latency': 138.303909559967, 'output_throughput': 62.174670458405025}
metrics['accuracy']=0.98

### CODE REVIEW

File: csrc/CMakeLists.txt [hunk 1]
```diff
@@ -78,6 +81,7 @@ target_link_libraries(${OP_PLUGIN_NAME} PRIVATE
         platform
         ascendalog
         dl
+        nnopbase
```
Comment (randgun) [inline csrc/CMakeLists.txt]: why you need it?

File: csrc/CMakeLists.txt [hunk 2]
```diff
@@ -91,6 +95,7 @@ target_include_directories(${OP_PLUGIN_NAME} PRIVATE
         ${TORCH_DIR}/include
         ${TORCH_DIR}/include/torch/csrc/api/include
         ${TORCH_NPU_DIR}/include
+        ${ASCEND_INCLUDE_DIR}/
```
Comment (randgun) [inline csrc/CMakeLists.txt]: why you need it?

File: csrc/sparse_flash_attention/op_host/sparse_flash_attention.cpp [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+// namespace scope global parameters
+uint32_t actualCaptureNum = 0;
+static std::unordered_map<uint64_t, uint32_t> captureMap;
+// at::Tensor workspace;
+
+namespace sglang {
+namespace npu_kernel {
+using namespace ge_helper;
+inline at::Tensor ConstructSparseFlashAttentionOutputTensor(const at::Tensor &query)
```
Comment (randgun) [inline csrc/sparse_flash_attention/op_host/sparse_flash_attention.cpp:45]: It's better to move these SFA specialized functions from npu_kernel namespace to SFAHost

File: csrc/sparse_flash_attention/op_host/sparse_flash_attention.cpp [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    if (layout_query.has_value()) {
+        layoutQuery = std::string(layout_query.value());
+        sfa.SetAttrStr("layout_query", layoutQuery);
+    }
+    if (layout_kv.has_value()) {
+        layoutKV = std::string(layout_kv.value());
+        sfa.SetAttrStr("layout_kv", layoutKV);
+    }
+
+    sfa.SetAttrAny("sparse_mode", static_cast<int64_t>(sparse_mode.value_or(3)));
```
Comment (randgun) [inline csrc/sparse_flash_attention/op_host/sparse_flash_attention.cpp:75]: You can get default value from def.h, instead of hard coding.

File: csrc/sparse_flash_attention/op_host/tiling/sparse_flash_attention_data.h [hunk 5]
```diff
@@ -0,0 +1,60 @@
+#ifndef SPARSE_FLASH_ATTENTION_DATA_H
+#define SPARSE_FLASH_ATTENTION_DATA_H
+
+#include <cstdint>
+
+#ifndef __CCE__
+#include "register/tilingdata_base.h"
```
Comment (randgun) [inline csrc/sparse_flash_attention/op_host/tiling/sparse_flash_attention_data.h]: It can be deleted, check it

File: csrc/sparse_flash_attention/op_kernel/sparse_flash_attention_kernel.cpp [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+                                                             GM_ADDR actualSeqLengthsQuery, GM_ADDR actualSeqLengthsKV,
+                                                             GM_ADDR queryRope, GM_ADDR keyRope, GM_ADDR attentionOut,
+                                                             GM_ADDR workspace, GM_ADDR tiling)
+{
+    KERNEL_TASK_TYPE_DEFAULT(KERNEL_TYPE_MIX_AIC_1_2);
+
+    TPipe tPipe;
+    __gm__ uint8_t *user = GetUserWorkspace(workspace);
+
+    // 1. 獲取 Tiling Data 並讀取 Key
```
Comment (randgun) [inline csrc/sparse_flash_attention/op_kernel/sparse_flash_attention_kernel.cpp]: comment in English

File: tests/python/sgl_kernel_npu/test_sparse_flash_attention.py [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+                mm2_res = np.matmul(softmax_res, v_sparse, dtype=matmul_dtype)
+                y[batch, n2_idx * g : (n2_idx + 1) * g, s1_idx, :] = mm2_res
+
+    if layout_query == "TND":
+        cpu_out = torch.tensor(y)
+        return trans_bnsd_to_tnd(cpu_out, cpu_out.shape, actual_seq_lengths_query)
+    return np.transpose(y, axes=(0, 2, 1, 3))
+
+
+class TestCustomSFA(unittest.TestCase):
```
Comment (randgun) [inline tests/python/sgl_kernel_npu/test_sparse_flash_attention.py:266]: Your test should cover bfloat16 type

File: csrc/sparse_flash_attention/op_kernel/sparse_flash_attention_common.h [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+
+    uint32_t bN2Start = 0U;
+    uint32_t bN2End = 0U;
+    uint32_t gS1Start = 0U;
+    uint32_t gS1End = 0U;
+
+    uint32_t tndFDCoreArrLen = 0U;      // TNDFlashDecoding相关分核信息array的长度
+    uint32_t coreStartKVSplitPos = 0U;  // TNDFlashDecoding kv起始位置
+
+    uint32_t mBaseSize = 1ULL;
```
Comment (randgun) [inline csrc/sparse_flash_attention/op_kernel/sparse_flash_attention_common.h:191]: Why use ULL?

Comment (AndyLi429) [reply csrc/sparse_flash_attention/op_kernel/sparse_flash_attention_common.h:191]: same as CANN recipes

File: csrc/sparse_flash_attention/op_host/sparse_flash_attention.cpp [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+        tilingData.splitKVParams.s2, tilingData.tilingKey);
+    auto hashValue = host_utils::TupleHasher::Hash(tup);
+    static auto globalTilingBuffer =
+        at::empty({tilingSize * MAX_CAPTURE_NUM}, at::TensorOptions().dtype(at::kByte).device(device));
+    at::Tensor tilingTensor;
+    if (captureMap.find(hashValue) != captureMap.end()) {
+        // For decode replay phase and part of prefill phase, get cached tiling data from globalTilingBuffer
+        tilingTensor = at::from_blob(globalTilingBuffer.data_ptr<uint8_t>() + (tilingSize * captureMap[hashValue]),
+                                     tilingSize, at::kByte);
+    } else if (actualCaptureNum >= MAX_CAPTURE_NUM / 2) {
```
Comment (RuixuanZhang06) [inline csrc/sparse_flash_attention/op_host/sparse_flash_attention.cpp]: MAX_CAPTURE_NUM / 2 =》 MAX_CAPTURE_NUM - Rerserved_tilling_buffer_size

File: csrc/sparse_flash_attention/README.md [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+
+# Call sparse flash attention
+scale_value = 1.0 / (head_dim ** 0.5)
+output = torch.ops.npu.sparse_flash_attention(
+    query, key, value, sparse_indices,
+    scale_value=scale_value,
+    sparse_block_size=sparse_block_size,
+    attention_out=attention_out,
+    layout_query='BSND',
+    layout_kv='BSND',
```
Comment (AndyLi429) [inline csrc/sparse_flash_attention/README.md]: comfirm

File: csrc/sparse_flash_attention/op_host/tiling/sparse_flash_attention_tiling.cpp [hunk 11]
```diff
@@ -0,0 +1,1791 @@
...
+
+    TORCH_CHECK(n2Size_ == 1,
+        OPS_LOG_E(opName_, "kv_head_num should be 1, but got %u", n2Size_));
+
+    TORCH_CHECK(n1Size_ % n2Size_ == 0,
+        OPS_LOG_E(opName_, "q_head_num(%u) must be divisible by kv_head_num(%u)", n1Size_, n2Size_));
+
+    std::vector<uint32_t> gSizeSupportList = {1, 2, 4, 8, 16, 32, 64, 128};
+    TORCH_CHECK(std::find(gSizeSupportList.begin(), gSizeSupportList.end(), gSize_) != gSizeSupportList.end(),
+        OPS_LOG_E(opName_, "group num should be in 1, 2, 4, 8, 16, 32, 64, 128, but got %u", gSize_));
```
[the remark this answers is not in the document]
Comment (AndyLi429) [reply csrc/sparse_flash_attention/op_host/tiling/sparse_flash_attention_tiling.cpp]: DONE

File: csrc/sparse_flash_attention/op_host/sparse_flash_attention.cpp (hunk 9 above)
Comment (AndyLi429) [reply csrc/sparse_flash_attention/op_host/sparse_flash_attention.cpp]: modify to original version, which is tested well

**Review by RuixuanZhang06 — DISMISSED** [review]
